### PR TITLE
Contract Revision for Approval

### PR DESCRIPTION
**Pull request by alliu930410**
- [x] Replaced tabs with double spaces in KlktnNFT.cdc contract
- [x] Add access(self) accessors to klktnNFTTypeSet, tokenExpiredPerType (removed in the 4th item below), tokenMintedPerType
- [x] Update flow.json contract path
- [x] Remove tokenExpiredPerType contract variable as suggested by the Flow team
- [x]  Add updateTemplateMetadata() mutator function and transaction
- [x] Add unit test for update_template_metadata.cdc transaction
- [x] Update NFTMinter resource to Admin resource

### CODE REVIEW

**Comment by alliu930410** [comment]
Reference: Comments from the Flow team

# cadence/contracts

# KlktnNFT.cdc

- [ ] 

* There are some tabs in the code, which otherwise uses two spaces for
indentation. This should be regularized.

32-40:
```
  // klktnNFTTypeSet:
  // - Dictionary for metaData and administrative parameters per typeID
  pub var klktnNFTTypeSet: {UInt64: KlktnNFTMetaData}
  // tokenExpiredPerType:
  // - Dictionary to hold expired token templates
  pub var tokenExpiredPerType: {UInt64: Bool}
  // tokenMintedPerType:
  // - Dictionary to track minted tokens per typeID
  pub var tokenMintedPerType: {UInt64: UInt64}
  ```

- [ ] 

The content of these dictionaries will be publicly writable by *any* user.

I recommend following TopShot's approach, making the member variables
`access(self)` and adding accessors.

If you wish the metadata to be updatable by admins, you will need to
add mutators in the Admin object.

https://github.com/dapperlabs/nba-smart-contracts/blob/master/contracts/TopShot.cdc#L99

https://github.com/dapperlabs/nba-smart-contracts/blob/master/contracts/TopShot.cdc#L700

263: `  pub fun checkTokenExpiration(typeID: UInt64): Bool {`

- [ ] 

Is there a particular reason for caching the expiry rather than
calculating it each time?

# cadence/transactions

## create_token_template.cdc

- [ ] awaiting for response

3: `// Question to Flow team: how to use flow CLI to pass in
dictionary {String: String} objects (the command aboce is not
working)?`

File: cadence/contracts/KlktnNFT.cdc [hunk 1]
```diff
@@ -199,22 +196,34 @@ pub contract KlktnNFT: NonFungibleToken {
...
-      // expire token when mintLimit is hit
-      if serialNumber >= targetTokenMetaData.mintLimit {
-        KlktnNFT.tokenExpiredPerType[typeID] = true
-      }
       // increse the serial number for the minted token type
       KlktnNFT.tokenMintedPerType[typeID] = serialNumber
-		}
+    }
+    
+    pub fun updateMetaData(typeID: UInt64, metaDataToUpdate: {String: String}): KlktnNFT.KlktnNFTMetaData {
```
Comment (flockonus) [inline cadence/contracts/KlktnNFT.cdc:212]: suggestion to make more explicit on function: `updateTemplateMetaData` (vs. updating a single NFT)

File: cadence/contracts/KlktnNFT.cdc [hunk 2]
```diff
@@ -199,22 +196,34 @@ pub contract KlktnNFT: NonFungibleToken {
...
+      if !KlktnNFT.klktnNFTTypeSet.containsKey(typeID) {
+        panic("Token with the typeID does not exist.")
+      }
+      // typeID cannot change
+      var NFTTemplateObj = KlktnNFT.klktnNFTTypeSet[typeID]!
+      let typeID = NFTTemplateObj.typeID
+      let tokenName = NFTTemplateObj.tokenName
+      let mintLimit = NFTTemplateObj.mintLimit
```
Comment (flockonus) [inline cadence/contracts/KlktnNFT.cdc:220]: 👍  good stuff we can't change these things

File: cadence/contracts/KlktnNFT.cdc [hunk 3]
```diff
@@ -31,13 +31,10 @@ pub contract KlktnNFT: NonFungibleToken {
   pub var totalSupply: UInt64
```
Comment (flockonus) [inline cadence/contracts/KlktnNFT.cdc:31]: Sorry this wasn't from the review, but following the same logic, can we confirm that totalSupply can't be altered, either by us or any other user?

Comment (flockonus) [reply cadence/contracts/KlktnNFT.cdc:31]: Same with `CollectionStoragePath` `CollectionStoragePath` `MinterStoragePath` ?

File: test/deploy.test.js [hunk 4]
```diff
@@ -209,7 +209,43 @@ describe("Test Transaction Code", () => {
...
+    try {
+      const updateMetaDataTxResult = await sendTransaction({ 
+        code: updateMetaDataTxTemplate,
+        args: [
+          [typeID, UInt64],
+          [
+            [
+            {key: 'artist', value: 'Kevin Woo'},
+            {key: 'releaseYear', value: '2021'},
+            {key: 'uri', value: 'ipfs://QmTv2Tx9XQeLrvg8rs9LCCih6FrHt2mXs3LVBt23ZD7eE7'},
```
Comment (flockonus) [inline test/deploy.test.js:230]: i'd suggest try if can we see these change to a value different than it was before?

File: cadence/contracts/KlktnNFT.cdc [hunk 5]
```diff
@@ -199,22 +196,34 @@ pub contract KlktnNFT: NonFungibleToken {
       // emit Minted event
       emit Minted(id: KlktnNFT.totalSupply, typeID: typeID, serialNumber: serialNumber, metaData: targetTokenMetaData.metaData)
```
Comment (flockonus) [inline cadence/contracts/KlktnNFT.cdc:197]: Unrelated to the review:
We're calling NFT `metaData` Topshot calls `metadata`

Maybe there is no mention of it in the NFT standard, but just want to be sure we're compatible with the spec?

**Review by flockonus — COMMENTED** [review]
@alliu930410 thanks for being on top of it so quickly!
Added comments, and some extra questions

File: test/deploy.test.js (hunk 4 above)
Comment (alliu930410) [reply test/deploy.test.js:229]: Added slightly different fields to test metadata update to existing and non-existing fields.

**Review by alliu930410 — COMMENTED** [review]
@flockonus I've updated 

- [x] naming convention metaData to metadata 
- [ ] NFTMinter resource to Admin resource (as it does more than just minting tokens and is aligned with NBA Topshots)
- [ ] the unit test to test update to both existing and non-exist metadata fields

File: cadence/contracts/KlktnNFT.cdc (hunk 1 above)
Comment (alliu930410) [reply cadence/contracts/KlktnNFT.cdc]: Updated the function name and corresponding transaction & unit tests

**Review by flockonus — APPROVED** [review]
Great work @alliu930410!
The access control of `pub` is a bit counter-intuitive to me, thanks for looking into that.
Feel free to send over to review again plese

**Pull request merged**